### PR TITLE
Fixed thread safety issues & allow cached OpenNLP models

### PR DESCRIPTION
**Pull request by dswitzer**
The `POSTaggerME` and `SentenceDetectorME` objects are not thread safe, so trying to re-use an instance of the `RakeAlgorithm` can throw exceptions or generate unreliable results.

I've added synchronize calls to the `getTokens()` function to make sure that calls can be safely made in multiple threads.

I also added constructors so that the cache instances of POSTaggerME & SentenceDetectorME and use those to construct an instance of `RakeAlgorithm`. I've updated the README.md with an example of this:

```java
import io.github.crew102.rapidrake.RakeAlgorithm;
import io.github.crew102.rapidrake.data.SmartWords;
import io.github.crew102.rapidrake.model.RakeParams;
import io.github.crew102.rapidrake.model.Result;
import io.github.crew102.rapidrake.opennlpUtils.Tagger;
import io.github.crew102.rapidrake.opennlpUtils.SentDetector;
import opennlp.tools.postag.POSTaggerME;
import opennlp.tools.sentdetect.SentenceDetectorME;

public class Example {

  public static void main(String[] args) throws java.io.IOException {
    
    // Create an object to hold algorithm parameters
    String[] stopWords = new SmartWords().getSmartWords(); 
    String[] stopPOS = {"VB", "VBD", "VBG", "VBN", "VBP", "VBZ"}; 
    int minWordChar = 1;
    boolean shouldStem = true;
    String phraseDelims = "[-,.?():;\"!/]"; 
    RakeParams params = new RakeParams(stopWords, stopPOS, minWordChar, shouldStem, phraseDelims);
    
    // Create a RakeAlgorithm object
    String POStaggerURL = "model-bin/en-pos-maxent.bin"; // The path to your POS tagging model
    String SentDetectURL = "model-bin/en-sent.bin"; // The path to your sentence detection model

    // create the required model classes, you can cache these instances in a singleton
    POSTaggerME tagger = new Tagger(POStaggerURL).getPosTagger();
    SentenceDetectorME sentDetect = new SentDetector(SentDetectURL).getSentDetector();

    // now creating an instance of the RakeAlgorithm is fast
    RakeAlgorithm rakeAlg = new RakeAlgorithm(params, tagger, sentDetect);
    
    // Call the rake method
    String txt = "dogs are great, don't you agree? I love dogs, especially big dogs";
    Result result = rakeAlg.rake(txt);
    
    // Print the resulting keywords (not stemmed)
    System.out.println(result.distinct());
    
  }
}

// [dogs (1.33), great (1), big dogs (3.33)]
```

### CODE REVIEW

**Comment by crew102** [comment]
Hi @dswitzer , thanks for the PR. A few questions:

> The POSTaggerME and SentenceDetectorME objects are not thread safe, so trying to re-use an instance of the RakeAlgorithm can throw exceptions or generate unreliable results.

Can you clarify the use case here? Maybe I'm misunderstanding what you mean by re-use, but shouldn't calling `rake()` on the same object multiple times (i.e., re-using the RakeAlgorithm instance) not be a problem insofar as multithreading is concerned, given that it wouldn't be multithreaded code by nature?


> I also added constructors so that the cache instances of POSTaggerME & SentenceDetectorME and use those to construct an instance of RakeAlgorithm. 

I'm also sorely out of date on this repo, so can you provide a reproducible example of how instantiating InputStream objects and using those in the constructor below would be slower than the method you provide?

https://github.com/crew102/rapidrake-java/blob/db5480a9d36ec89ff07597d32ddeed87d1b8d190/src/main/java/io/github/crew102/rapidrake/RakeAlgorithm.java#L62-L66

**Comment by dswitzer** [comment]
@crew102,

In my use case, I'm trying to use the `rake()` method to extract keywords from strings. This is part of a workflow which is triggered to provide live feedback to the user, so I need to call `rake()` frequently. The problem I was running into is creating a new instance of the `RakeAlgorithm` is too expensive, due to the overhead in loading the OpenNLP models. It's only taking 500-600ms, but for this use case, that is way to high (because I'm trying to get sub 200ms responses for the entire workflow and extracting keywords is just one part of the process).

The way I'm trying to solve that problem is by caching the instance in a singleton layer, but that is where my problems cropped up.


> Hi @dswitzer , thanks for the PR. A few questions:
> 
> > The POSTaggerME and SentenceDetectorME objects are not thread safe, so trying to re-use an instance of the RakeAlgorithm can throw exceptions or generate unreliable results.
> 
> Can you clarify the use case here? Maybe I'm misunderstanding what you mean by re-use, but shouldn't calling `rake()` on the same object multiple times (i.e., re-using the RakeAlgorithm instance) not be a problem insofar as multithreading is concerned, given that it wouldn't be multithreaded code by nature?

So the thread safety issues in the `RakeAlgorithm.getTokens()` and are due to issues in the OpenNLP. Neither the `POSTaggerME.tag()` nor the `SentenceDetectorME.sentDetect()` are safe to use in concurrent operations. Each running JVM thread needs it's own instance or you need to synchronize the calls, otherwise you run into errors. 

Synchronizing the blocks of code were we make the calls to the libraries was just the easiest solution.

Without the `synchronized` blocks around `SentenceDetectorME.sentDetect()` and `POSTaggerME.tag()` calls, when my code was running under load, I'd start getting random errors and java.lang.NullPointerException errors.

> > I also added constructors so that the cache instances of POSTaggerME & SentenceDetectorME and use those to construct an instance of RakeAlgorithm.
> 
> I'm also sorely out of date on this repo, so can you provide a reproducible example of how instantiating InputStream objects and using those in the constructor below would be slower than the method you provide?
> 
> https://github.com/crew102/rapidrake-java/blob/db5480a9d36ec89ff07597d32ddeed87d1b8d190/src/main/java/io/github/crew102/rapidrake/RakeAlgorithm.java#L62-L66

The problem is with the performance of `getPosTagger()` and `getSentDetector()`. That's where the 500-600ms overhead in initializing an instance `RakeAlgorithm` was coming into play. By creating and caching the `tagger` and `sentDetector` instances, I can then just create a new instance of `RakeAlgorithm` with basically no overhead.

So the `RakeAlgorithm(RakeParams rakeParams, Tagger taggerInstance, SentDetector sentDectInstance)` really is going to have the same basic performance as using the path or InputStream constructors. However, if create instances of the POSTaggerME and SentenceDetectorME and cache those and then use the `RakeAlgorithm(RakeParams rakeParams, POSTaggerME posTaggerMEInstance, SentenceDetectorME sentDetectorMEInstance)` constructor, you do not have that initial performance hit.

However, because those instances are not thread safe when used in concurrency, it's why the code needed the `synchronized` blocks.

So in summary, caching the `POSTaggerME` and `SentenceDetectorME` objects saves me 500-600ms on each run (since I'm using already loaded models), while the `synchronized` blocks allow the code to run safely when used concurrently.

You could probably squeeze more performance out by caching instances of `POSTaggerME` and `SentenceDetectorME` per thread, but the synchronized blocks seem to run fast enough for my purposes.

File: src/main/java/io/github/crew102/rapidrake/RakeAlgorithm.java [hunk 1]
```diff
@@ -64,6 +64,36 @@ public RakeAlgorithm(RakeParams rakeParams, InputStream taggerStream, InputStrea
...
+  /**
+   * Constructor.
+   *
+   * @param rakeParams the parameters RAKE will use
+   * @param taggerInstance An instance of io.github.crew102.rapidrake.opennlpUtils.Tagger.
+   * @param sentDectInstance An instance of io.github.crew102.rapidrake.opennlpUtils.SentDetector.
+   * @see RakeParams
+   * @throws java.io.IOException if either of the input streams are invalid
+   */
+  public RakeAlgorithm(RakeParams rakeParams, Tagger taggerInstance, SentDetector sentDectInstance) throws java.io.IOException {
```
Comment (crew102) [inline src/main/java/io/github/crew102/rapidrake/RakeAlgorithm.java]: What's the use case for this constructor, given the `getPosTagger()` call is what takes the time?

**Comment by crew102** [comment]
K I think I'm mostly following. A few follow up questions.

* Presumably you need to be updating the RakeParams as your user interacts with the app, and that's why you need to create new instances of the RakeAlgorithm? Otherwise you could just call rake on the various strings that the user passes you to, all using the same object, similar to how is done in this test:

https://github.com/crew102/rapidrake-java/blob/6bfd081cf05c5dcbb05e3e4dee9bca7bc4e46afe/src/test/java/io/github/crew102/rapidrake/TestRapidRake.java#L25-L32

...Or maybe the pattern you're using is related to the multithreaded nature of your app?

* I agree that the first constructor you have implemented (passing in posTaggerMEInstance and sentDetectorMEInstance) makes sense. Having trouble understanding the use case for the second one though.

* Question about how you're using multithreading - What exactly are you parallelizing across threads? In other words, at what point in your app do you split things along threads? Is it like a document-based parallelization, similar to what is described in the last section on this post: https://crew102.github.io/slowraker/articles/rapidraker.html#making-rapidrake-even-faster. Or have you refactored the underlying rapidrake-java code to split work across threads within the algorithm itself?

**Comment by dswitzer** [comment]
That constructor really doesn’t offer anything. I implemented that
originally to see if that offered any benefits, but it does not.

I should have just removed it.

**Comment by dswitzer** [comment]
RE: Question about how you're using multithreading - What exactly are you
parallelizing across threads? In other words, at what point in your app do
you split things along threads?

This is a web app, so there are many users using an instance of the
RakeAlgorith at one time. It’s cached in a singleton, so can be accessed by
many threads at once.

Since the OpenNLP methods being used are not safe to be used in multiple
threads, that’s why I added the synchronized blocks.

If you take the first unit test you reference and try running the following
line concurrently across a bunch of different threads, you should start
seeing exceptions being thrown in 1.14:

Result res2 = minRakeAlg.rake("also, here is some more. text.");

Under 1.15 it should work fine.

**Comment by crew102** [comment]
Cool, if you're good with the refactorings I put on top of your PR (see https://github.com/crew102/rapidrake-java/tree/feature/pr-7-refactorings), we're good to go. You can pull those two commits into this PR and I'll merge. Thanks again.

File: src/main/java/io/github/crew102/rapidrake/RakeAlgorithm.java (hunk 1 above)
Comment (dswitzer) [reply src/main/java/io/github/crew102/rapidrake/RakeAlgorithm.java]: That constructor really doesn’t offer anything. I implemented that
originally to see if that offered any benefits, but it does not.

I should have just removed it.

**Comment by dswitzer** [comment]
@crew102:

I've merged in the changes, so hopefully this is ready to go.
